[Mac Catalyst] Adjust autocorrection underline color based on caret color

pxlcoder opened the pull request:
#### b199f7db6bc8daa14464c22f44e3d70845e791f7
<pre>
[Mac Catalyst] Adjust autocorrection underline color based on caret color
<a href="https://bugs.webkit.org/show_bug.cgi?id=256475">https://bugs.webkit.org/show_bug.cgi?id=256475</a>
rdar://108355409

Reviewed by Wenson Hsieh.

The autocorrection underline color should be a variant of the current caret
color. In order to facilitate this, the insertion point color is plumbed from
the UI Process into the Web Process.

* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::autocorrectionReplacementMarkerColor const):
(WebCore::RenderTheme::documentMarkerLineColor const):
* Source/WebCore/rendering/RenderTheme.h:

Add a `RenderText` parameter to `documentMarkerLineColor`, since it is
necessary to determine the caret color.

* Source/WebCore/rendering/RenderThemeCocoa.h:
* Source/WebCore/rendering/RenderThemeCocoa.mm:
(WebCore::RenderThemeCocoa::platformAutocorrectionReplacementMarkerColor const): Deleted.

Remove override, as the color is too dynamic to be cached. `caret-color` can
differ per-element, so the autocorrection underline color can no longer be cached.

* Source/WebCore/rendering/RenderThemeIOS.h:

Add a static method to set the insertion point color to avoid unnecessary
initialization of the singleton.

* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::cachedInsertionPointColor):
(WebCore::RenderThemeIOS::insertionPointColor):
(WebCore::RenderThemeIOS::autocorrectionReplacementMarkerColor const):

Adjust the autocorrection underline color based on the current caret color. If
`caret-color` is `auto`, use the default color from the UI process. The
adjustment is performed using the HSL colorspace, matching UIKit.

(WebCore::RenderThemeIOS::setInsertionPointColor):
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter&lt;TextBoxPath&gt;::paintPlatformDocumentMarker):
* Source/WebKit/Shared/WebPageCreationParameters.cpp:
(WebKit::WebPageCreationParameters::encode const):
(WebKit::WebPageCreationParameters::decode):
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _insertionPointColor]):

Get the insertion point color from the text input traits if it exists. Otherwise,
use the default color specified in UIKit.

* Source/WebKit/UIProcess/PageClient.h:

Add a `PageClient` hook to get the insertion point color.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::insertionPointColor):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView tintColorDidChange]):

Ensure the insertion point color is up-to-date, following changes to the view&apos;s
tint color.

* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::insertionPointColorDidChange):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::platformInitialize):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::setInsertionPointColor):

Canonical link: <a href="https://commits.webkit.org/263879@main">https://commits.webkit.org/263879@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/c346ce9507c4d13d43146cb3c82af29dd86caacf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/5817 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/5985 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/6173 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/7377 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/6208 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/5818 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/6207 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/5952 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/8001 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/5923 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/5971 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/5270 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/7432 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/3448 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/5246 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/13208 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/5316 "Passed tests") | [❌ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/5325 "5 failures") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/7538 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/5770 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/4739 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/5211 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/9332 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/701 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/5572 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

whsieh commented on Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
```diff
@@ -647,6 +647,14 @@ - (void)_updateScrollViewIndicatorStyle
+    if (auto *insertionPointColor = [[self _textInputTraits] insertionPointColor])
+        return insertionPointColor;
+
+    return UIColor.insertionPointColor;
```
Nit - possibly a bit more concise as `return self._textInputTraits.insertionPointColor ?: UIColor.insertionPointColor;`.

whsieh commented on Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
```diff
@@ -1078,6 +1078,11 @@ static UIInterfaceOrientationMask toUIInterfaceOrientationMask(WebCore::ScreenOr
     return color;
 }
 
+WebCore::Color PageClientImpl::insertionPointColor()
```
Nit - you can omit the WebCore:: here and below, since this file already uses `namespace WebCore;`.

whsieh commented on Source/WebCore/rendering/RenderThemeIOS.mm:
```diff
@@ -1358,6 +1364,36 @@ static bool shouldUseConvexGradient(const Color& backgroundColor)
+
+Color RenderThemeIOS::autocorrectionReplacementMarkerColor(const RenderText& renderer, OptionSet<StyleColorOptions>) const
+{
+    Color caretColor = CaretBase::computeCaretColor(renderer.style(), renderer.textNode());
```
```suggestion
    auto caretColor = CaretBase::computeCaretColor(renderer.style(), renderer.textNode());
```

whsieh commented on Source/WebCore/rendering/RenderThemeIOS.mm:
```diff
@@ -1358,6 +1364,36 @@ static bool shouldUseConvexGradient(const Color& backgroundColor)
+    return *cachedInsertionPointColor();
+}
+
+Color RenderThemeIOS::autocorrectionReplacementMarkerColor(const RenderText& renderer, OptionSet<StyleColorOptions>) const
```
Nit - would it be cleaner to just pass in a `renderer` here, and use `renderer.styleColorOptions()` to grab the color options from it? (So that the function never takes a set of `StyleColorOptions` that doesn't match the `renderer`).

pxlcoder replied:
I was already using `renderer.styleColorOptions()` in the method body – removed the parameter.

aprotyas commented on Source/WebCore/rendering/RenderTheme.cpp:
```diff
@@ -2041,7 +2041,7 @@ Color RenderTheme::platformDictationAlternativesMarkerColor(OptionSet<StyleColor
 }
 
-Color RenderTheme::autocorrectionReplacementMarkerColor(OptionSet<StyleColorOptions> options) const
+Color RenderTheme::autocorrectionReplacementMarkerColor(const RenderText&, OptionSet<StyleColorOptions> options) const
```
Nit: Since `options == RenderText::styleColorOptions()`, do we need the `options` argument?

pxlcoder replied:
Removed!

pxlcoder committed:
[Mac Catalyst] Adjust autocorrection underline color based on caret color
https://bugs.webkit.org/show_bug.cgi?id=256475
rdar://108355409

Reviewed by Wenson Hsieh.

The autocorrection underline color should be a variant of the current caret
color. In order to facilitate this, the insertion point color is plumbed from
the UI Process into the Web Process.

* Source/WebCore/rendering/RenderTheme.cpp:
(WebCore::RenderTheme::autocorrectionReplacementMarkerColor const):
(WebCore::RenderTheme::documentMarkerLineColor const):
* Source/WebCore/rendering/RenderTheme.h:

Add a `RenderText` parameter to `documentMarkerLineColor`, since it is
necessary to determine the caret color.

* Source/WebCore/rendering/RenderThemeCocoa.h:
* Source/WebCore/rendering/RenderThemeCocoa.mm:
(WebCore::RenderThemeCocoa::platformAutocorrectionReplacementMarkerColor const): Deleted.

Remove override, as the color is too dynamic to be cached. `caret-color` can
differ per-element, so the autocorrection underline color can no longer be cached.

* Source/WebCore/rendering/RenderThemeIOS.h:

Add a static method to set the insertion point color to avoid unnecessary
initialization of the singleton.

* Source/WebCore/rendering/RenderThemeIOS.mm:
(WebCore::cachedInsertionPointColor):
(WebCore::RenderThemeIOS::insertionPointColor):
(WebCore::RenderThemeIOS::autocorrectionReplacementMarkerColor const):

Adjust the autocorrection underline color based on the current caret color. If
`caret-color` is `auto`, use the default color from the UI process. The
adjustment is performed using the HSL colorspace, matching UIKit.

(WebCore::RenderThemeIOS::setInsertionPointColor):
* Source/WebCore/rendering/TextBoxPainter.cpp:
(WebCore::TextBoxPainter<TextBoxPath>::paintPlatformDocumentMarker):
* Source/WebKit/Shared/WebPageCreationParameters.cpp:
(WebKit::WebPageCreationParameters::encode const):
(WebKit::WebPageCreationParameters::decode):
* Source/WebKit/Shared/WebPageCreationParameters.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.h:
* Source/WebKit/UIProcess/API/ios/WKWebViewIOS.mm:
(-[WKWebView _insertionPointColor]):

Get the insertion point color from the text input traits if it exists. Otherwise,
use the default color specified in UIKit.

* Source/WebKit/UIProcess/PageClient.h:

Add a `PageClient` hook to get the insertion point color.

* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::creationParameters):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.h:
* Source/WebKit/UIProcess/ios/PageClientImplIOS.mm:
(WebKit::PageClientImpl::insertionPointColor):
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView tintColorDidChange]):

Ensure the insertion point color is up-to-date, following changes to the view's
tint color.

* Source/WebKit/UIProcess/ios/WebPageProxyIOS.mm:
(WebKit::WebPageProxy::insertionPointColorDidChange):
* Source/WebKit/WebProcess/WebPage/Cocoa/WebPageCocoa.mm:
(WebKit::WebPage::platformInitialize):
* Source/WebKit/WebProcess/WebPage/WebPage.h:
* Source/WebKit/WebProcess/WebPage/WebPage.messages.in:
* Source/WebKit/WebProcess/WebPage/ios/WebPageIOS.mm:
(WebKit::WebPage::setInsertionPointColor):

Canonical link: https://commits.webkit.org/263879@main
